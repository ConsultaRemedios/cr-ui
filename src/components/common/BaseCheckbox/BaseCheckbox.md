### Basic Usage

BaseCheckbox component follows the one way data flow, it means the component does not have an
internal state to control if it is checked or not. The **checked** prop passed to it is the only
one responsible for handling the checkbox state. So, if you don't update the state that is passed
as the checked prop, the BaseCheckbox will be stay freezed. See the examples:

#### Checked
```js
<BaseCheckbox name="newsletter" value="on" label="This is a checked checkbox" :checked="true"/>
```

#### Unchecked
```js
<BaseCheckbox name="newsletter" value="on" label="This is a unchecked checkbox"/>
```

#### Disabled
```js
<BaseCheckbox name="newsletter" value="" label="This is a disable checkbox" :disabled="true"/>
```

#### Update checked prop using the one way data flow
```js
new Vue({
  data() {
    return {
      newsletter: '',
    };
  },
  methods: {
    onChangeCheckbox({ name, value }) {
      this[name] = value;
    },
    isChecked(name) {
      return this[name] === 'on';
    },
  },
  template: `
    <div>
      <BaseCheckbox
        name="newsletter"
        value="on"
        label="This is a checked checkbox"
        :checked="isChecked('newsletter')"
        @change="onChangeCheckbox"
      />
    </div>
  `,
})
```