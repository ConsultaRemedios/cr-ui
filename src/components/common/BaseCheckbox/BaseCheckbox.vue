<template>
  <div>
    <label :class="[
      $style.wrapper,
      { [$style.isSelected]: checked, [$style.isDisabled]: disabled }
    ]">
      <input
        :class="$style.input"
        :name="name"
        type="checkbox"
        :checked="checked"
        :disabled="disabled"
        v-bind="$attrs"
        @change="onChange"
      >

      <span
        v-if="label"
        :class="$style.label"
      >
        {{ label }}
      </span>
    </label>
  </div>
</template>

<script>
export default {
  name: 'BaseCheckbox',
  inheritAttrs: false,

  props: {
    checked: {
      type: Boolean,
      default: false,
    },

    disabled: {
      type: Boolean,
      default: false,
    },
    value: {
      type: [String, Number],
      required: true,
    },
    name: {
      type: String,
      required: true,
    },
    label: {
      type: String,
      default: '',
    },

    model: {
      type: String,
      default: '',
    },
  },

  methods: {
    onChange() {
      this.$emit('change', {
        name: this.model || this.name,
        value: this.checked ? '' : this.value,
      });
    },
  },
};
</script>

<style module>
  .wrapper {
    cursor: pointer;
    display: flex;
    position: relative;
    font-weight: 400;
  }

  .wrapper::before {
    content: '';
    flex-shrink: 0;
    display: inline-block;
    width: 20px;
    height: 20px;
    border: 1px solid #BDBDBD;
    border-radius: 3px;
    margin-right: 10px;
  }

  .wrapper:hover::before {
    border: 1px solid #138484;
  }

  .wrapper.isSelected::before {
    border-color: #138484;
    background-color: #138484;
  }

  .wrapper.isDisabled::before {
    background-color: #F2F2F2;
    border: 2px solid #DADADA;
  }

  .wrapper::after {
    content: '';
    display: block;
    width: 16px;
    height: 12px;
    background: url('./../../../assets/check-icon.svg');
    position: absolute;
    top: 4px;
    left: 2px;

    transform: scale(0);
    transition: transform 0.1s ease-out;
  }

  .label {
    line-height: 1.4;
  }

  .wrapper.isSelected::after {
    transform: scale(1);
  }

  .wrapper.isDisabled::after {
    background: #DADADA;
    width: 12px;
    height: 4px;
    top: 8px;
    left: 4px;
    transform: scale(1);
  }

  .input {
    display: none;
  }
</style>
