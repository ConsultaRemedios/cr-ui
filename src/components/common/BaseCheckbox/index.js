import BaseCheckbox from './BaseCheckbox.vue';

export default BaseCheckbox;
