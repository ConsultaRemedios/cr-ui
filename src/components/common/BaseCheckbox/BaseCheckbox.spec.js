import { mount } from '@vue/test-utils';

import BaseCheckbox from './BaseCheckbox.vue';

describe('BaseCheckbox Component', () => {
  const defaultProps = {
    name: 'newsletter',
    value: 'on',
  };

  const mountBaseCheckbox = (props = {}, attributes = {} ) => {
    const propsData = { ...defaultProps, ...props };
    const attrs = { ...attributes };
    return mount(BaseCheckbox, { propsData, attrs });
  };

  describe('Defaults', () => {
    let wrapper;

    beforeEach(() => {
      wrapper = mountBaseCheckbox();
    });

    it('assigns checked prop as false', () => {
      expect(wrapper.vm.checked).toBe(false);
    });
  });

  describe('Behavior', () => {
    describe('When it is checked and a click happens', () => {
      describe('and model prop is empty', () => {
        it('emmits "change" event with the proper params', () => {
          const wrapper = mountBaseCheckbox({ checked: true });
          wrapper.find('input').trigger('change');
          const event = wrapper.emitted().change[0][0];
          expect(event).toMatchObject({ name: 'newsletter', value: '' });
        });
      });

      describe('and model prop is not empty', () => {
        it('emmits "change" event with the proper params', () => {
          const wrapper = mountBaseCheckbox({ checked: true, model: 'customModel' });
          wrapper.find('input').trigger('change');

          const event = wrapper.emitted().change[0][0];

          expect(event).toMatchObject({ name: 'customModel', value: '' });
        });
      });
    });

    describe('When it is not checked and a click happens', () => {
      describe('and model prop is empty', () => {
        it('emmits "change" event with the proper params', () => {
          const wrapper = mountBaseCheckbox({ checked: false });
          wrapper.find('input').trigger('change');
          const event = wrapper.emitted().change[0][0];

          expect(event).toMatchObject({ name: 'newsletter', value: 'on' });
        });
      });

      describe('and model prop is not empty', () => {
        it('emmits "change" event with the proper params', () => {
          const wrapper = mountBaseCheckbox({ checked: false, model: 'customModel' });
          wrapper.find('input').trigger('change');
          const event = wrapper.emitted().change[0][0];

          expect(event).toMatchObject({ name: 'customModel', value: 'on' });
        });
      });
    });

    describe('When it is disabled and a click happens', () => {
      it('it does not call onChange method', () => {
        const wrapper = mountBaseCheckbox({ disabled: true });
        wrapper.setMethods({ onChange: jest.fn() });
        wrapper.find('input').trigger('click');
        expect(wrapper.vm.onChange).not.toBeCalled();
      });
    });
  });

  describe('Render', () => {
    describe('When it is disabled', () => {
      it('matches the snapshot', () => {
        const wrapper = mountBaseCheckbox({ disabled: true});
        expect(wrapper.element).toMatchSnapshot();
      });
    });
    describe('When it is not checked', () => {
      it('matches the snapshot', () => {
        const wrapper = mountBaseCheckbox({ checked: true });
        expect(wrapper.element).toMatchSnapshot();
      })
    });

    describe('When it is checked', () => {
      it('matches the snapshot', () => {
        const wrapper = mountBaseCheckbox({ checked: false });
        expect(wrapper.element).toMatchSnapshot();
      })
    });

    describe('When there is label', () => {
      it('renders the snapshot', () => {
        const wrapper = mountBaseCheckbox({ label: 'Newsletter' });
        expect(wrapper.element).toMatchSnapshot();
        expect(wrapper.vm.label).toBe('Newsletter');
      });
    });

    describe('When there is no label', () => {
      it('renders the snapshot', () => {
        const wrapper = mountBaseCheckbox();
        expect(wrapper.element).toMatchSnapshot();
      });
    });

    describe('When there is id', () => {
      it('renders the snapshot', () => {
        const wrapper = mountBaseCheckbox(null, { id: 'checkbox-term' });
        expect(wrapper.element).toMatchSnapshot();
        expect(wrapper.vm.$attrs.id).toBe('checkbox-term');
      });
    });

    describe('When there is no id', () => {
      it('renders the snapshot', () => {
        const wrapper = mountBaseCheckbox();
        expect(wrapper.element).toMatchSnapshot();
      });
    });
  });
});
