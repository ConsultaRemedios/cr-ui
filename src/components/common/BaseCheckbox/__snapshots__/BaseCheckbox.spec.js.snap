// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BaseCheckbox Component Render When it is checked matches the snapshot 1`] = `
<div>
  <label
    class="wrapper"
  >
    <input
      class="input"
      name="newsletter"
      type="checkbox"
    />
     
    <!---->
  </label>
</div>
`;

exports[`BaseCheckbox Component Render When it is disabled matches the snapshot 1`] = `
<div>
  <label
    class="wrapper isDisabled"
  >
    <input
      class="input"
      disabled="disabled"
      name="newsletter"
      type="checkbox"
    />
     
    <!---->
  </label>
</div>
`;

exports[`BaseCheckbox Component Render When it is not checked matches the snapshot 1`] = `
<div>
  <label
    class="wrapper isSelected"
  >
    <input
      class="input"
      name="newsletter"
      type="checkbox"
    />
     
    <!---->
  </label>
</div>
`;

exports[`BaseCheckbox Component Render When there is id renders the snapshot 1`] = `
<div>
  <label
    class="wrapper"
  >
    <input
      class="input"
      id="checkbox-term"
      name="newsletter"
      type="checkbox"
    />
     
    <!---->
  </label>
</div>
`;

exports[`BaseCheckbox Component Render When there is label renders the snapshot 1`] = `
<div>
  <label
    class="wrapper"
  >
    <input
      class="input"
      name="newsletter"
      type="checkbox"
    />
     
    <span
      class="label"
    >
      
      Newsletter
    
    </span>
  </label>
</div>
`;

exports[`BaseCheckbox Component Render When there is no id renders the snapshot 1`] = `
<div>
  <label
    class="wrapper"
  >
    <input
      class="input"
      name="newsletter"
      type="checkbox"
    />
     
    <!---->
  </label>
</div>
`;

exports[`BaseCheckbox Component Render When there is no label renders the snapshot 1`] = `
<div>
  <label
    class="wrapper"
  >
    <input
      class="input"
      name="newsletter"
      type="checkbox"
    />
     
    <!---->
  </label>
</div>
`;
